src/net_processing.cpp

```diff
@@ -2295,7 +2295,9 @@ bool PeerManagerImpl::AlreadyHaveTx(const GenTxid& gtxid, bool include_reconside
 
     const uint256& hash = gtxid.GetHash();
 
-    if (m_orphanage.HaveTx(gtxid)) return true;
+    // Orphanage is checked by wtxid. However, even if this is a txid, look up the same hash in
+    // case this is a non-segwit transaction in the orphanage.
+    if (m_orphanage.HaveTx(Wtxid::FromUint256(gtxid.GetHash()))) return true;
```

Comment:
It's not immediately obvious to me why it is safe to just convert any `gtxid` into a `Wtxid` here, especially since there are callsites of `AlreadyHaveTx` where a ` GenTxId::Txid` is passed, such as e.g. [here](https://github.com/bitcoin/bitcoin/blob/d73245abc70346a0e8805d50a1f395706084294c/src/net_processing.cpp#L4632). If it is indeed safe, I think it could be a better approach to update `AlreadyHaveTx` to take a `Wtxid` and push the conversion further to the edge, so it can be reviewed on a case-by-case basis?

Reply:
(Ignore me if I'm understanding, still picking things up 😅) -- coinbase transactions are [required to have all `0` Wtxids](https://github.com/bitcoin/bips/blob/master/bip-0141.mediawiki#commitment-structure) (`0x0000....0000`) which means for an `INV` message for a witness coinbase you have to use the non-witness txid to reference the transaction? - so `AlreadyHaveTx` in that case is explicitly *not* using a Wtxid? (Or are coinbases never `INV`'d?)

Also perhaps not great/confusing to treat Wtxid and Txids as always interchangeable (though the types allow it) because of the coinbase Wtxid caveat?

Reply:
> It's not immediately obvious to me why it is safe to just convert any gtxid into a Wtxid here

The argument I'm making in this PR is that in all the places we are checking by txid, we shouldn't be doing that, because we're missing same-txid-different-witness cases. But yeah totally fair to say it's difficult to review it together, so I'll split the first commit to make it more explicit where the change is happening and why that's ok.

>  I think it could be a better approach to update AlreadyHaveTx to take a Wtxid

fwiw I disagree, as we do want to check txids in the other data structures.

> coinbase transactions are [required to have all 0 Wtxids](https://github.com/bitcoin/bips/blob/master/bip-0141.mediawiki#commitment-structure) (0x0000....0000) which means for an INV message for a witness coinbase you have to use the non-witness txid to reference the transaction? - so AlreadyHaveTx in that case is explicitly not using a Wtxid? (Or are coinbases never INV'd?)

That link is referring to the wtxid of the coinbase transaction when calculating the witness commitment; that's not the case in any other situation. Also, yes, if we get a coinbase transaction in tx relay we'd drop that pretty much immediately and wouldn't be putting it in orphanage:
https://github.com/bitcoin/bitcoin/blob/9d1a286f20b8a602ffe72928bcd79be09fdbf9d0/src/validation.cpp#L735-L742

Reply:
> The argument I'm making in this PR is that in all the places we are checking by txid, we shouldn't be doing that, because we're missing same-txid-different-witness cases.

I agree, the rationale makes sense. My concern is about segwit txs (`txid != wtxid`) where we check by txid (for whatever reason: intentional because we don't have witness data, legacy and we never updated it, buggy, ...): just creating a wtxid from the txid hash as is done here seems like it has potential to start missing things we previously wouldn't - since `TxOrphanage` now _only_ indexes by `wtxid`?

I've made a branch where I've created 2 overloads of `AlreadyHaveTx`: one that takes `Wtxid` and one that takes `Txid`, to force all the conversions to the edges. I implemented the `Txid` to perform less checks, based on the documentation (didn't verify the code) that `m_recent_rejects_reconsiderable` and `m_recent_rejects` uses `wtxid`, whereas `m_recent_confirmed_transactions` uses both `txid` and `wtxid`.

There are 4 callsites of `AlreadyHaveTx`. For 2 of them, [1) here](https://github.com/glozow/bitcoin/compare/2024-04-orphan-use-wtxid...stickies-v:bitcoin:pr/30000#diff-6875de769e90cec84d2e8a9c1b962cdbcda44d870d42e4215827e599e11e90e3R4255-R4256) and [2) here](https://github.com/glozow/bitcoin/compare/2024-04-orphan-use-wtxid...stickies-v:bitcoin:pr/30000#diff-6875de769e90cec84d2e8a9c1b962cdbcda44d870d42e4215827e599e11e90e3R4561), the conversion seems trivial/sensible.

A third is more confusing to me. [3) Here in `ProcessMessage`](https://github.com/glozow/bitcoin/compare/2024-04-orphan-use-wtxid...stickies-v:bitcoin:pr/30000#diff-6875de769e90cec84d2e8a9c1b962cdbcda44d870d42e4215827e599e11e90e3R4663), I initially used the `Txid` overload, but that fails the `p2p_orphan_handling.py::test_orphan_of_orphan` test. Switching to `Wtxid` makes it pass, but I don't understand why. Since with this pull, `TxOrphanage` indexes on `Wtxid`, calling `HaveTx` on a blind conversion of `parent_txid` into a `Wtxid` seems like it shouldn't find anything in the orphanage, unless we're dealing with non-segwit transaction?

Finally, for the [4) fourth here](https://github.com/glozow/bitcoin/compare/2024-04-orphan-use-wtxid...stickies-v:bitcoin:pr/30000#diff-6875de769e90cec84d2e8a9c1b962cdbcda44d870d42e4215827e599e11e90e3R6319-R6341), I think my refactoring is a no-op, we just blindly call the different function signatures based on `GenTxid::IsWtxid()`, no real difference with the current implementation except perhaps for making things more explicit.

Reply:
To be clear: I'm not saying this PR needs to overhaul `AlreadyHaveTx` like I did, and I think having `TxOrphanage` index on `Wtxid` instead of `Txid` is better. I just think that if we're going to be calling `TxOrphanage` methods with (what are in fact) txids, we should probably have `TxOrphanage` _also_ keep a `Txid` index and not to blind conversion such as is happening here? I think that would address my concerns and make things easier to review?

Reply:
fwiw, I think the in-orphanage txid==wtixd case seems pretty rare all things considered. I had 4 over a 24 hour period, vs about 400 times where we fetched a segwit parent via txid though that particular txid was in the orphanage already.

Reply:
> iow I'm not sure extra complexity is worth it,

Oh, fair enough. I don't have a good enough view of the attack vectors here. In my review, I was operating on the assumption that we wouldn't want to miss anything we're currently doing. If it's okay to relax that, that could make sense - but it should probably be made explicit then?

Reply:
IIUC If the check is removed, you'll waste bandwidth when you get orphan chains that are non-segwit, but you won't keep or relay them, so it shouldn't churn the orphanage.

Would it help if the comment was touched up to be explicit about what it's saving us from doing?

Reply:
imo `wtxid` and `txid` being double SHA256 outputs and `wtxid` being essentially a superset of `txid`, makes this blind conversion legitimate.

Reply:
Split up the commit, hope it helps (https://github.com/bitcoin/bitcoin/pull/30000#pullrequestreview-2049866191)